How do I get multiple return values from the contract function?

There is a function of such a contract.
`function getValues() public constant returns (uint[], byte[64][], address[])`

I wrote the Rust code to receive the value, but this is an error.
```
let result = contract.query("getValues", (), None, Options::default());
let values: Vec<Token> = result.wait().unwrap();
```
```
error[E0277]: the trait bound `std::vec::Vec<ethabi::Token>: web3::contract::tokens::Tokenizable` is not satisfied
```

It worked when you made changes to crate, but I think this is a bad way.

`use ethabi::Token;` -> `use web3::contract::tokens::Token;`

> rust-web3/src/contract/tokens.rs:3
`use ethabi::Token;` -> `pub use ethabi::Token;`

> ethabi/src/token/token.rs:7
`#[derive(Debug, PartialEq, Clone)]` -> `#[derive(Debug, PartialEq, Clone, Deserialize)]`

How is the best way to do it?
Thank you.

Thank you for advice.
I tried it, but it was an error.

```
error[E0277]: the trait bound `(std::vec::Vec<web3::types::U256>, std::vec::Vec<std::vec::Vec<u8>>, std::vec::Vec<web3::types::H160>): web3::contract::tokens::Tokenizable` is not satisfied
   --> src\test.rs:236:71
    |
236 |             contract.query("getValues", (), None, Options::default()).wait().unwrap();
    |                                                                       ^^^^ the trait `web3::contract::tokens::Tokenizable` is not implemented for `(std::vec::Vec<web3::types::U256>, std::vec::Vec<std::vec::Vec<u8>>, std::vec::Vec<web3::types::H160>)`
    |
    = note: required because of the requirements on the impl of `web3::contract::tokens::Output` for `(std::vec::Vec<web3::types::U256>, std::vec::Vec<std::vec::Vec<u8>>, std::vec::Vec<web3::types::H160>)`
    = note: required because of the requirements on the impl of `jsonrpc_core::<unnamed>::Future` for `web3::contract::result::QueryResult<(std::vec::Vec<web3::types::U256>, std::vec::Vec<std::vec::Vec<u8>>, std::vec::Vec<web3::types::H160>), web3::transports::http::FetchTask>`
```

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -10,6 +10,7 @@ keywords = ["jsonrpc", "web3", "ethereum", "rpc", "client"]
 authors = ["Tomasz Drwięga <tomasz@ethcore.io>"]
 
 [dependencies]
+arrayvec = "0.3"
 ethabi = "2.0"
 futures = "0.1"
 jsonrpc-core = "7.0"

--- a/src/contract/mod.rs
+++ b/src/contract/mod.rs
@@ -4,7 +4,7 @@ use ethabi;
 
 use api::Eth;
 use contract::result::QueryResult;
-use contract::tokens::{Output, Tokens};
+use contract::tokens::{Deserialize, Serialize};
 use types::{Address, Bytes, CallRequest, H256, TransactionRequest, TransactionCondition, U256};
 use {Transport, Error as ApiError};
 
@@ -85,7 +85,7 @@ impl<T: Transport> Contract<T> {
 
   /// Execute a contract function
   pub fn call<P>(&self, func: &str, params: P, from: Address, options: Options) -> QueryResult<H256, T::Out> where
-    P: Tokens,
+    P: Serialize,
   {
     self.abi.function(func.into())
       .and_then(|function| function.encode_call(params.into_tokens()))
@@ -107,7 +107,7 @@ impl<T: Transport> Contract<T> {
 
   /// Estimate gas required for this function call.
   pub fn estimate_gas<P>(&self, func: &str, params: P, from: Address, options: Options) -> QueryResult<U256, T::Out> where
-    P: Tokens,
+    P: Serialize,
   {
     self.abi.function(func.into())
       .and_then(|function| function.encode_call(params.into_tokens()))
@@ -127,9 +127,9 @@ impl<T: Transport> Contract<T> {
 
   /// Call constant function
   pub fn query<R, A, P>(&self, func: &str, params: P, from: A, options: Options) -> QueryResult<R, T::Out> where
-    R: Output,
+    R: Deserialize,
     A: Into<Option<Address>>,
-    P: Tokens,
+    P: Serialize,
   {
     self.abi.function(func.into())
       .and_then(|function| function.encode_call(params.into_tokens()).map(|call| (call, function)))

--- a/src/contract/mod.rs
+++ b/src/contract/mod.rs
@@ -4,7 +4,7 @@ use ethabi;
 
 use api::Eth;
 use contract::result::QueryResult;
-use contract::tokens::{Deserialize, Serialize};
+use contract::tokens::{Detokenize, Tokenize};
 use types::{Address, Bytes, CallRequest, H256, TransactionRequest, TransactionCondition, U256};
 use {Transport, Error as ApiError};
 
@@ -85,7 +85,7 @@ impl<T: Transport> Contract<T> {
 
   /// Execute a contract function
   pub fn call<P>(&self, func: &str, params: P, from: Address, options: Options) -> QueryResult<H256, T::Out> where
-    P: Serialize,
+    P: Tokenize,
   {
     self.abi.function(func.into())
       .and_then(|function| function.encode_call(params.into_tokens()))
@@ -107,7 +107,7 @@ impl<T: Transport> Contract<T> {
 
   /// Estimate gas required for this function call.
   pub fn estimate_gas<P>(&self, func: &str, params: P, from: Address, options: Options) -> QueryResult<U256, T::Out> where
-    P: Serialize,
+    P: Tokenize,
   {
     self.abi.function(func.into())
       .and_then(|function| function.encode_call(params.into_tokens()))
@@ -127,9 +127,9 @@ impl<T: Transport> Contract<T> {
 
   /// Call constant function
   pub fn query<R, A, P>(&self, func: &str, params: P, from: A, options: Options) -> QueryResult<R, T::Out> where
-    R: Deserialize,
+    R: Detokenize,
     A: Into<Option<Address>>,
-    P: Serialize,
+    P: Tokenize,
   {
     self.abi.function(func.into())
       .and_then(|function| function.encode_call(params.into_tokens()).map(|call| (call, function)))

--- a/src/contract/result.rs
+++ b/src/contract/result.rs
@@ -5,7 +5,7 @@ use futures::{Future, Async, Poll};
 use serde;
 
 use contract;
-use contract::tokens::Deserialize;
+use contract::tokens::Detokenize;
 use helpers;
 use rpc;
 use types::Bytes;
@@ -63,7 +63,7 @@ impl<T, F> QueryResult<T, F> where
   }
 }
 
-impl<T: Deserialize + serde::de::DeserializeOwned, F> Future for QueryResult<T, F> where
+impl<T: Detokenize + serde::de::DeserializeOwned, F> Future for QueryResult<T, F> where
   F: Future<Item=rpc::Value, Error=ApiError>,
 {
   type Item = T;

--- a/src/contract/result.rs
+++ b/src/contract/result.rs
@@ -5,7 +5,7 @@ use futures::{Future, Async, Poll};
 use serde;
 
 use contract;
-use contract::tokens::Output;
+use contract::tokens::Deserialize;
 use helpers;
 use rpc;
 use types::Bytes;
@@ -63,7 +63,7 @@ impl<T, F> QueryResult<T, F> where
   }
 }
 
-impl<T: Output + serde::de::DeserializeOwned, F> Future for QueryResult<T, F> where
+impl<T: Deserialize + serde::de::DeserializeOwned, F> Future for QueryResult<T, F> where
   F: Future<Item=rpc::Value, Error=ApiError>,
 {
   type Item = T;

--- a/src/contract/tokens.rs
+++ b/src/contract/tokens.rs
@@ -6,12 +6,12 @@ use contract::Error;
 use types::{self, Address, H256, U256};
 
 /// Output type possible to deserialize from Contract ABI
-pub trait Output {
+pub trait Deserialize {
   /// Creates a new instance from parsed ABI tokens.
   fn from_tokens(tokens: Vec<Token>) -> Result<Self, Error> where Self: Sized;
 }
 
-impl<T: Tokenizable> Output for T {
+impl<T: Tokenizable> Deserialize for T {
   fn from_tokens(mut tokens: Vec<Token>) -> Result<Self, Error> {
     if tokens.len() != 1 {
       return Err(Error::InvalidOutputType(format!("Expected single element, got a list: {:?}", tokens)));
@@ -22,7 +22,7 @@ impl<T: Tokenizable> Output for T {
 
 macro_rules! impl_output {
   ($num: expr, $( $ty: ident : $no: tt, )+) => {
-    impl<$($ty, )+> Output for ($($ty,)+) where
+    impl<$($ty, )+> Deserialize for ($($ty,)+) where
       $(
         $ty: Tokenizable,
       )+
@@ -51,26 +51,26 @@ impl_output!(4, A:0, B:1, C:2, D:3, );
 impl_output!(5, A:0, B:1, C:2, D:3, E:4, );
 
 /// Tokens conversion trait
-pub trait Tokens {
+pub trait Serialize {
   /// Convert to list of tokens
   fn into_tokens(self) -> Vec<Token>;
 }
 
-impl<'a> Tokens for &'a [Token] {
+impl<'a> Serialize for &'a [Token] {
   fn into_tokens(self) -> Vec<Token> { self.to_vec() }
 }
 
-impl<T: Tokenizable> Tokens for T {
+impl<T: Tokenizable> Serialize for T {
   fn into_tokens(self) -> Vec<Token> { vec![self.into_token()] }
 }
 
-impl Tokens for () {
+impl Serialize for () {
   fn into_tokens(self) -> Vec<Token> { vec![] }
 }
 
 macro_rules! impl_tokens {
   ($( $ty: ident : $no: tt, )+) => {
-    impl<$($ty, )+> Tokens for ($($ty,)+) where
+    impl<$($ty, )+> Serialize for ($($ty,)+) where
       $(
         $ty: Tokenizable,
       )+
@@ -255,10 +255,10 @@ impl_fixed_array!(1024);
 #[cfg(test)]
 mod tests {
   use ethabi::Token;
-  use super::Output;
+  use super::Deserialize;
   use types::{Address, U256};
 
-  fn output<R: Output>() -> R {
+  fn output<R: Deserialize>() -> R {
     unimplemented!()
   }
 

--- a/src/contract/tokens.rs
+++ b/src/contract/tokens.rs
@@ -1,5 +1,6 @@
 //! Contract Functions Output types.
 
+use arrayvec::ArrayVec;
 use ethabi::Token;
 use contract::Error;
 use types::{self, Address, H256, U256};
@@ -10,11 +11,44 @@ pub trait Output {
   fn from_tokens(tokens: Vec<Token>) -> Result<Self, Error> where Self: Sized;
 }
 
-impl Output for Vec<Token> {
-  fn from_tokens(tokens: Vec<Token>) -> Result<Self, Error> {
-    Ok(tokens)
+impl<T: Tokenizable> Output for T {
+  fn from_tokens(mut tokens: Vec<Token>) -> Result<Self, Error> {
+    if tokens.len() != 1 {
+      return Err(Error::InvalidOutputType(format!("Expected single element, got a list: {:?}", tokens)));
+    }
+    Self::from_token(tokens.drain(..).next().expect("At least one element in vector; qed"))
   }
 }
+
+macro_rules! impl_output {
+  ($num: expr, $( $ty: ident : $no: tt, )+) => {
+    impl<$($ty, )+> Output for ($($ty,)+) where
+      $(
+        $ty: Tokenizable,
+      )+
+    {
+      fn from_tokens(mut tokens: Vec<Token>) -> Result<Self, Error> {
+        if tokens.len() != $num {
+          return Err(Error::InvalidOutputType(format!(
+            "Expected {} elements, got a list of {}: {:?}",
+            $num,
+            tokens.len(),
+            tokens
+          )));
+        }
+        let mut it = tokens.drain(..);
+        Ok(($(
+          $ty::from_token(it.next().expect("All elements are in vector; qed"))?,
+        )+))
+      }
+    }
+  }
+}
+
+impl_output!(2, A:0, B:1, );
+impl_output!(3, A:0, B:1, C:2, );
+impl_output!(4, A:0, B:1, C:2, D:3, );
+impl_output!(5, A:0, B:1, C:2, D:3, E:4, );
 
 /// Tokens conversion trait
 pub trait Tokens {
@@ -62,15 +96,6 @@ pub trait Tokenizable {
   fn from_token(token: Token) -> Result<Self, Error> where Self: Sized;
   /// Converts a specified type back into token.
   fn into_token(self) -> Token;
-}
-
-impl<T: Tokenizable> Output for T {
-  fn from_tokens(mut tokens: Vec<Token>) -> Result<Self, Error> {
-    if tokens.len() != 1 {
-      return Err(Error::InvalidOutputType(format!("Expected single element, got a list: {:?}", tokens)));
-    }
-    Self::from_token(tokens.drain(..).next().expect("At least one element in vector; qed"))
-  }
 }
 
 impl Tokenizable for Token {
@@ -134,7 +159,7 @@ impl Tokenizable for Address {
 impl Tokenizable for U256 {
   fn from_token(token: Token) -> Result<Self, Error> {
     match token {
-      Token::Uint(data) => Ok(U256(data)),
+      Token::Int(data) | Token::Uint(data) => Ok(U256(data)),
       other => Err(Error::InvalidOutputType(format!("Expected `U256`, got {:?}", other))),
     }
   }
@@ -160,10 +185,97 @@ impl Tokenizable for Vec<u8> {
   fn from_token(token: Token) -> Result<Self, Error> {
     match token {
       Token::Bytes(data) => Ok(data),
-      other => Err(Error::InvalidOutputType(format!("Expected `bool`, got {:?}", other))),
+      other => Err(Error::InvalidOutputType(format!("Expected `bytes`, got {:?}", other))),
     }
   }
   fn into_token(self) -> Token {
     Token::Bytes(self)
+  }
+}
+
+impl<T: Tokenizable> Tokenizable for Vec<T> {
+  fn from_token(token: Token) -> Result<Self, Error> {
+    match token {
+      Token::FixedArray(tokens) | Token::Array(tokens) => tokens.into_iter().map(Tokenizable::from_token).collect(),
+      other => Err(Error::InvalidOutputType(format!("Expected `Array`, got {:?}", other))),
+    }
+  }
+
+  fn into_token(self) -> Token {
+    Token::Array(self.into_iter().map(Tokenizable::into_token).collect())
+  }
+}
+
+macro_rules! impl_fixed_array {
+  ($num: expr) => {
+    impl<T: Tokenizable + Clone> Tokenizable for [T; $num] {
+      fn from_token(token: Token) -> Result<Self, Error> {
+        match token {
+          Token::FixedArray(tokens) => {
+            if tokens.len() != $num {
+              return Err(Error::InvalidOutputType(format!("Expected `FixedArray({})`, got FixedArray({})", $num, tokens.len())));
+            }
+
+            let mut arr = ArrayVec::<[T; $num]>::new();
+            let mut it = tokens.into_iter().map(T::from_token);
+            for _ in 0..$num {
+              arr.push(it.next().expect("Length validated in guard; qed")?);
+            }
+            // Can't use expect here because [T; $num]: Debug is not satisfied.
+            match arr.into_inner() {
+              Ok(arr) => Ok(arr),
+              Err(_) => panic!("All elements inserted so the array is full; qed"),
+            }
+          },
+          other => Err(Error::InvalidOutputType(format!("Expected `FixedArray({})`, got {:?}", $num, other))),
+        }
+      }
+
+      fn into_token(self) -> Token {
+        Token::FixedArray(ArrayVec::from(self).into_iter().map(T::into_token).collect())
+      }
+    }
+  }
+}
+
+impl_fixed_array!(1);
+impl_fixed_array!(2);
+impl_fixed_array!(3);
+impl_fixed_array!(4);
+impl_fixed_array!(5);
+impl_fixed_array!(8);
+impl_fixed_array!(16);
+impl_fixed_array!(32);
+impl_fixed_array!(64);
+impl_fixed_array!(128);
+impl_fixed_array!(256);
+impl_fixed_array!(512);
+impl_fixed_array!(1024);
+
+#[cfg(test)]
+mod tests {
+  use ethabi::Token;
+  use super::Output;
+  use types::{Address, U256};
+
+  fn output<R: Output>() -> R {
+    unimplemented!()
+  }
+
+  #[test]
+  #[ignore]
+  fn should_be_able_to_compile() {
+    let _tokens: Vec<Token> = output();
+    let _uint: U256 = output();
+    let _address: Address = output();
+    let _string: String = output();
+    let _bool: bool = output();
+    let _bytes: Vec<u8> = output();
+
+    let _pair: (U256, bool) = output();
+    let _vec: Vec<U256> = output();
+    let _array: [U256; 4] = output();
+
+    let _mixed: (Vec<Vec<u8>>, [U256; 4], Vec<U256>, U256) = output();
   }
 }

--- a/src/contract/tokens.rs
+++ b/src/contract/tokens.rs
@@ -6,12 +6,12 @@ use contract::Error;
 use types::{self, Address, H256, U256};
 
 /// Output type possible to deserialize from Contract ABI
-pub trait Deserialize {
+pub trait Detokenize {
   /// Creates a new instance from parsed ABI tokens.
   fn from_tokens(tokens: Vec<Token>) -> Result<Self, Error> where Self: Sized;
 }
 
-impl<T: Tokenizable> Deserialize for T {
+impl<T: Tokenizable> Detokenize for T {
   fn from_tokens(mut tokens: Vec<Token>) -> Result<Self, Error> {
     if tokens.len() != 1 {
       return Err(Error::InvalidOutputType(format!("Expected single element, got a list: {:?}", tokens)));
@@ -22,7 +22,7 @@ impl<T: Tokenizable> Deserialize for T {
 
 macro_rules! impl_output {
   ($num: expr, $( $ty: ident , )+) => {
-    impl<$($ty, )+> Deserialize for ($($ty,)+) where
+    impl<$($ty, )+> Detokenize for ($($ty,)+) where
       $(
         $ty: Tokenizable,
       )+
@@ -52,26 +52,26 @@ impl_output!(4, A, B, C, D, );
 impl_output!(5, A, B, C, D, E, );
 
 /// Tokens conversion trait
-pub trait Serialize {
+pub trait Tokenize {
   /// Convert to list of tokens
   fn into_tokens(self) -> Vec<Token>;
 }
 
-impl<'a> Serialize for &'a [Token] {
+impl<'a> Tokenize for &'a [Token] {
   fn into_tokens(self) -> Vec<Token> { self.to_vec() }
 }
 
-impl<T: Tokenizable> Serialize for T {
+impl<T: Tokenizable> Tokenize for T {
   fn into_tokens(self) -> Vec<Token> { vec![self.into_token()] }
 }
 
-impl Serialize for () {
+impl Tokenize for () {
   fn into_tokens(self) -> Vec<Token> { vec![] }
 }
 
 macro_rules! impl_tokens {
   ($( $ty: ident : $no: tt, )+) => {
-    impl<$($ty, )+> Serialize for ($($ty,)+) where
+    impl<$($ty, )+> Tokenize for ($($ty,)+) where
       $(
         $ty: Tokenizable,
       )+
@@ -256,10 +256,10 @@ impl_fixed_array!(1024);
 #[cfg(test)]
 mod tests {
   use ethabi::Token;
-  use super::Deserialize;
+  use super::Detokenize;
   use types::{Address, U256};
 
-  fn output<R: Deserialize>() -> R {
+  fn output<R: Detokenize>() -> R {
     unimplemented!()
   }
 

--- a/src/contract/tokens.rs
+++ b/src/contract/tokens.rs
@@ -21,7 +21,7 @@ impl<T: Tokenizable> Deserialize for T {
 }
 
 macro_rules! impl_output {
-  ($num: expr, $( $ty: ident : $no: tt, )+) => {
+  ($num: expr, $( $ty: ident , )+) => {
     impl<$($ty, )+> Deserialize for ($($ty,)+) where
       $(
         $ty: Tokenizable,
@@ -45,10 +45,11 @@ macro_rules! impl_output {
   }
 }
 
-impl_output!(2, A:0, B:1, );
-impl_output!(3, A:0, B:1, C:2, );
-impl_output!(4, A:0, B:1, C:2, D:3, );
-impl_output!(5, A:0, B:1, C:2, D:3, E:4, );
+impl_output!(1, A, );
+impl_output!(2, A, B, );
+impl_output!(3, A, B, C, );
+impl_output!(4, A, B, C, D, );
+impl_output!(5, A, B, C, D, E, );
 
 /// Tokens conversion trait
 pub trait Serialize {

--- a/src/lib.rs
+++ b/src/lib.rs
@@ -2,6 +2,7 @@
 
 #![warn(missing_docs)]
 
+extern crate arrayvec;
 extern crate ethabi;
 extern crate jsonrpc_core as rpc;
 extern crate parking_lot;

